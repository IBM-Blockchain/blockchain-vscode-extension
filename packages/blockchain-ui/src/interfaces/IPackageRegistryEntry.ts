interface IPackageRegistryEntry {
    name: string;
    path: string;
    version?: string;
    sizeKB: number;
}

export default IPackageRegistryEntry;
