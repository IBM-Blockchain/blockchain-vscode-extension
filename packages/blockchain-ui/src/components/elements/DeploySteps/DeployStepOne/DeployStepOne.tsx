import React, { Component } from 'react';
import { Dropdown, Accordion, AccordionItem, InlineNotification, NotificationActionButton } from 'carbon-components-react';
import IPackageRegistryEntry from '../../../../interfaces/IPackageRegistryEntry';

interface IProps {
    workspaceNames: string[];
    deletedSelectedPackage: boolean;
    packageEntries: IPackageRegistryEntry[];
    selectedWorkspace: string | undefined;
    selectedPackage: IPackageRegistryEntry | undefined; // Package the user has selected in dropdown
    onPackageChange: (selectedPackage: IPackageRegistryEntry | undefined, workspaceName?: string) => void; // Callback to DeployPage
    onPackageWorkspace: (workspaceName: string) => void;
}

interface StepOneState {
    workspaceNames: string[];
    packageEntries: IPackageRegistryEntry[];
}

class DeployStepOne extends Component<IProps, StepOneState> {

    unselectedPackageLabel: string = 'Select smart contract';

    constructor(props: Readonly<IProps>) {
        super(props);
        this.state = {
            packageEntries: this.props.packageEntries,
            workspaceNames: this.props.workspaceNames.length > 0 ? this.props.workspaceNames : []
        };

        this.formatAllEntries = this.formatAllEntries.bind(this);
        this.formatPackageEntry = this.formatPackageEntry.bind(this);
        this.formatWorkspaceEntry = this.formatWorkspaceEntry.bind(this);
        this.selectPackage = this.selectPackage.bind(this);
        this.packageWorkspace = this.packageWorkspace.bind(this);

    }

    componentWillReceiveProps(receivedProps: IProps): void {

        let receivedNewPackages: boolean = false;

        if (receivedProps.packageEntries.length === this.state.packageEntries.length) {
            // Check each individual entry
            for (let i: number = 0; i < receivedProps.packageEntries.length; i++) {
                const newProp: IPackageRegistryEntry = receivedProps.packageEntries[i];
                const currentState: IPackageRegistryEntry = this.state.packageEntries[i];
                if (newProp.name !== currentState.name || newProp.version !== currentState.version) {
                    receivedNewPackages = true;
                    break;
                }
            }
        } else {
            // Different length, so different packages
            receivedNewPackages = true;
        }

        if (receivedNewPackages) {
            this.setState({packageEntries: receivedProps.packageEntries});
        }
    }

    selectPackage(data: any): void {

        // TODO: At some point we should move PackageRegistry to common, and call 'get' from here.
        // There would also be no need to pass in all entries when creating DeployView.

        let packageName: string;
        let packageVersion: string | undefined;

        if (data.selectedItem.includes(' (packaged)')) {
            // If selected a package

             // Get 'name@version' or just 'name'
            const packageIdentifier: string = data.selectedItem.split(' (packaged')[0];
            if (packageIdentifier.includes('@')) {
                // If it has a version
                const _splitParts: string[] = packageIdentifier.split('@');
                packageName = _splitParts[0];
                packageVersion = _splitParts[1];
            } else {
                // If it doesn't have a version
                packageName = packageIdentifier;
            }

            const selectedPackage: IPackageRegistryEntry | undefined = this.state.packageEntries.find((_entry: IPackageRegistryEntry) => {
                return (packageVersion && _entry.name === packageName && _entry.version === packageVersion) || (!packageVersion && _entry.name === packageName);
            }) as IPackageRegistryEntry;

            this.props.onPackageChange(selectedPackage);
        } else {
            // If selected an open workspace
            this.props.onPackageChange(undefined, data.selectedItem.split(' (open project)')[0]);

        }

    }

    formatAllEntries(): string[] {
        const packageNames: string[] = [];
        for (const name of this.state.workspaceNames) {
            const entryName: string = this.formatWorkspaceEntry(name);
            packageNames.push(entryName);
        }

        for (const entry of this.state.packageEntries) {
            const entryName: string = this.formatPackageEntry(entry);
            packageNames.push(entryName);
        }

        return packageNames;
    }

    formatPackageEntry(entry: IPackageRegistryEntry): string {
        if (entry.version) {
            return `${entry.name}@${entry.version} (packaged)`;
        } else {
            return `${entry.name} (packaged)`;
        }
    }

    formatWorkspaceEntry(workspaceName: string): string {
        return `${workspaceName} (open project)`;
    }

    packageWorkspace(): void {
        this.props.onPackageWorkspace(this.props.selectedWorkspace as string);
    }

    render(): JSX.Element {
        const items: string[] = this.formatAllEntries();

        let deletedMessage: JSX.Element | undefined;
        let packageMessage: JSX.Element | undefined;
        let packageStillExists: boolean = false;

        if (this.props.selectedPackage) {

            const packageToCheck: IPackageRegistryEntry | undefined = this.state.packageEntries.find((_package: IPackageRegistryEntry) => {
                return _package.name === (this.props.selectedPackage as IPackageRegistryEntry).name && _package.version === (this.props.selectedPackage as IPackageRegistryEntry).version;
            });

            packageStillExists = !!packageToCheck;

            if (!packageStillExists) {
                // User deleted the selected package.
                this.props.onPackageChange(undefined);
            }
        } else {
            if (this.props.deletedSelectedPackage) {
                // If the selected package was deleted whilst in Step Two or Three
                deletedMessage = (
                    <div className='bx--row margin-bottom-03'>
                        <div className='bx--col-lg-10'>
                            <InlineNotification
                            lowContrast={true}
                            kind='warning'
                            hideCloseButton={true}
                            title='The package you selected has been deleted'
                            />
                        </div>
                    </div>
                );
            }

            if (this.props.selectedWorkspace) {

                packageMessage = (
                    <div className='bx--row margin-bottom-03'>
                        <div className='bx--col-lg-10'>
                            <InlineNotification
                                actions={<NotificationActionButton style={{fontWeight: 600}} onClick={this.packageWorkspace}>Package open project</NotificationActionButton>}
                                hideCloseButton={true}
                                kind='info'
                                lowContrast={true}
                                notificationType='inline'
                                role='alert'
                                statusIconDescription='describes the status icon'
                                subtitle='before it can be used.'
                                title='The contract must be packaged'
                            />
                        </div>
                    </div>
                );

            }
        }

        return(
            <>
                <div className={'bx--row' + (packageMessage || deletedMessage ? '' : ' margin-bottom-10')}>
                    <div className='bx--col'>
                        <div style={{width: 384}}>
                            <Dropdown
                                ariaLabel='dropdown'
                                id='package-select'
                                invalidText='A valid value is required'
                                items={items}
                                label={this.unselectedPackageLabel}
                                titleText='Choose a smart contract to deploy'
                                type='default'
                                selectedItem={this.props.selectedPackage && packageStillExists  ? this.formatPackageEntry(this.props.selectedPackage) : (this.props.selectedWorkspace ? this.formatWorkspaceEntry(this.props.selectedWorkspace) :  this.unselectedPackageLabel)}
                                onChange={this.selectPackage}
                            />
                        </div>
                    </div>
                </div>
                {deletedMessage}
                {packageMessage}
                <div className='bx--row'>
                    <div className='bx--col-lg-10'>
                        <Accordion>
                            <AccordionItem title={'How does Fabric v2.X smart contract deployment work?'}>
                                <p>TODO</p>
                            </AccordionItem>
                        </Accordion>
                    </div>
                </div>
            </>
        );
    }
}

export default DeployStepOne;
