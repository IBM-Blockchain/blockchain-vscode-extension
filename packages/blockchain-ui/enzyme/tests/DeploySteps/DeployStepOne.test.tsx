import React from 'react';
import renderer from 'react-test-renderer';
import chai from 'chai';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';
import { ReactWrapper, mount, shallow, ShallowWrapper } from 'enzyme';

import DeployStepOne from '../../../src/components/elements/DeploySteps/DeployStepOne/DeployStepOne';
import IPackageRegistryEntry from '../../../src/interfaces/IPackageRegistryEntry';

chai.should();
chai.use(sinonChai);

// tslint:disable no-unused-expression

describe('DeployStepOne component', () => {
    let mySandBox: sinon.SinonSandbox;

    let onPackageChangeStub: sinon.SinonStub;
    let onPackageWorkspaceStub: sinon.SinonStub;
    let packageOne: IPackageRegistryEntry;
    let packageTwo: IPackageRegistryEntry;
    let packageThree: IPackageRegistryEntry;

    beforeEach(async () => {
        mySandBox = sinon.createSandbox();

        onPackageChangeStub = mySandBox.stub();
        onPackageWorkspaceStub = mySandBox.stub();
        packageOne = {name: 'mycontract', version: '0.0.1', path: '/package/one', sizeKB: 9000};
        packageTwo = {name: 'othercontract', version: '0.0.2', path: '/package/two', sizeKB: 12000};
        packageThree = {name: 'importedContract', path: '/package/two', sizeKB: 18000};
    });

    afterEach(async () => {
        mySandBox.restore();
    });

    describe('render', () => {
        it('should render the expected snapshot', () => {
            const component: any = renderer
                .create(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={packageOne} selectedWorkspace={undefined} workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />)
                .toJSON();
            expect(component).toMatchSnapshot();
        });

        it('should call use default chosen contract if no package passed through', () => {
            mount(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={undefined} selectedWorkspace={undefined} workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            onPackageChangeStub.should.not.have.been.calledWith(undefined);

        });

        it('should call onPackageChange if packge no longer exists', () => {
            mount(<DeployStepOne packageEntries={[packageTwo, packageThree]} selectedPackage={packageOne} selectedWorkspace={undefined} workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            onPackageChangeStub.should.have.been.calledOnceWith(undefined);
        });

        it('should select package if it still exists', () => {
            mount(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={packageOne} selectedWorkspace={undefined} workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            onPackageChangeStub.should.not.have.been.calledWith(undefined);
        });

        it('should display warning that package has been deleted, whilst in step two or three', () => {
            const component: ShallowWrapper<DeployStepOne> = shallow(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={undefined} selectedWorkspace={undefined} workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={true} onPackageChange={onPackageChangeStub} />);
            component.html().includes(`The package you selected has been deleted`).should.equal(true);
        });

        it('should display message that workspace must be packaged before it can be deployed', () => {
            const component: ShallowWrapper<DeployStepOne> = shallow(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={undefined} selectedWorkspace='workspaceOne' workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            component.html().includes(`The contract must be packaged`).should.equal(true);
        });

        it('should handle no workspaces being passed in', () => {
            const component: ShallowWrapper<DeployStepOne> = shallow(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={packageOne} selectedWorkspace={undefined} workspaceNames={[]} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            component.html().includes(`(open project)`).should.equal(false);
        });

    });

    describe('componentWillReceiveProps', () => {

        it('should update packages if size of new list is different', () => {
            const component: ReactWrapper<DeployStepOne> = mount(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={packageOne} selectedWorkspace={undefined} workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            const instance: DeployStepOne = component.instance() as DeployStepOne;

            const setStateStub: sinon.SinonStub = mySandBox.stub(instance, 'setState');

            instance.componentWillReceiveProps({
                packageEntries: [packageTwo, packageThree],
                selectedPackage: packageOne,
                onPackageChange: onPackageChangeStub,
                deletedSelectedPackage: false,
                workspaceNames: ['workspaceOne'],
                selectedWorkspace: undefined,
                onPackageWorkspace: onPackageWorkspaceStub
            });

            setStateStub.should.have.been.calledOnceWith({packageEntries: [packageTwo, packageThree]});
        });

        it('should update packages if the name of an entry is different', () => {
            const component: ReactWrapper<DeployStepOne> = mount(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={packageOne} selectedWorkspace={undefined} workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            const instance: DeployStepOne = component.instance() as DeployStepOne;

            const setStateStub: sinon.SinonStub = mySandBox.stub(instance, 'setState');

            const newPackage: IPackageRegistryEntry = {name: 'newPackage', version: '0.0.2', path: '/other/path', sizeKB: 4000};

            instance.componentWillReceiveProps({
                packageEntries: [packageOne, newPackage, packageThree],
                selectedPackage: packageOne,
                onPackageChange: onPackageChangeStub,
                deletedSelectedPackage: false,
                workspaceNames: ['workspaceOne'],
                selectedWorkspace: undefined,
                onPackageWorkspace: onPackageWorkspaceStub
            });

            setStateStub.should.have.been.calledOnceWith({packageEntries: [packageOne, newPackage, packageThree]});
        });

        it('should update packages if the version of an entry is different', () => {
            const component: ReactWrapper<DeployStepOne> = mount(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={packageOne} selectedWorkspace={undefined} workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            const instance: DeployStepOne = component.instance() as DeployStepOne;

            const setStateStub: sinon.SinonStub = mySandBox.stub(instance, 'setState');

            const newPackage: IPackageRegistryEntry = {name: 'packageTwo', version: '0.0.3', path: '/other/path', sizeKB: 4000};

            instance.componentWillReceiveProps({
                packageEntries: [packageOne, newPackage, packageThree],
                selectedPackage: packageOne,
                onPackageChange: onPackageChangeStub,
                deletedSelectedPackage: false,
                workspaceNames: ['workspaceOne'],
                selectedWorkspace: undefined,
                onPackageWorkspace: onPackageWorkspaceStub
            });

            setStateStub.should.have.been.calledOnceWith({packageEntries: [packageOne, newPackage, packageThree]});
        });

        it('should do nothing if received packages are the same', () => {
            const component: ReactWrapper<DeployStepOne> = mount(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={packageOne} selectedWorkspace={undefined} workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            const instance: DeployStepOne = component.instance() as DeployStepOne;

            const setStateStub: sinon.SinonStub = mySandBox.stub(instance, 'setState');

            instance.componentWillReceiveProps({
                packageEntries: [packageOne, packageTwo, packageThree],
                selectedPackage: packageOne,
                onPackageChange: onPackageChangeStub,
                deletedSelectedPackage: false,
                workspaceNames: ['workspaceOne'],
                selectedWorkspace: undefined,
                onPackageWorkspace: onPackageWorkspaceStub
            });

            setStateStub.should.not.have.been.calledWith({packageEntries: [packageOne, packageTwo, packageThree]});
        });
    });

    describe('formatAllEntries', () => {
        it('should format all entries to strings', () => {
            const component: ReactWrapper<DeployStepOne> = mount(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedWorkspace={undefined} workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} selectedPackage={packageOne} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            const instance: DeployStepOne = component.instance() as DeployStepOne;
            const packageEntryStrings: string[] = instance.formatAllEntries();

            packageEntryStrings.should.deep.equal([`workspaceOne (open project)`, `${packageOne.name}@${packageOne.version} (packaged)`, `${packageTwo.name}@${packageTwo.version} (packaged)`, `${packageThree.name} (packaged)`]);
        });
    });

    describe('formatPackageEntry', () => {
        it('should format package with version', () => {
            const component: ReactWrapper<DeployStepOne> = mount(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={packageOne} selectedWorkspace={undefined} workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            const instance: DeployStepOne = component.instance() as DeployStepOne;
            const packageString: string = instance.formatPackageEntry(packageOne);
            packageString.should.equal(`${packageOne.name}@${packageOne.version} (packaged)`);
        });

        it('should format package without version', () => {
            const component: ReactWrapper<DeployStepOne> = mount(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={packageOne} selectedWorkspace={undefined} workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            const instance: DeployStepOne = component.instance() as DeployStepOne;
            const packageString: string = instance.formatPackageEntry(packageThree);
            packageString.should.equal(`${packageThree.name} (packaged)`);
        });

    });

    describe('formatWorkspaceEntry', () => {
        it('should format workspace entry', () => {
            const component: ReactWrapper<DeployStepOne> = mount(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={packageOne} selectedWorkspace={undefined} workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            const instance: DeployStepOne = component.instance() as DeployStepOne;
            const workspaceString: string = instance.formatWorkspaceEntry('workspaceOne');
            workspaceString.should.equal(`workspaceOne (open project)`);
        });

    });

    describe('selectPackage', () => {
        it('should be able to select a package with a version', () => {
            const component: ReactWrapper<DeployStepOne> = mount(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={packageOne} selectedWorkspace={undefined} workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            const instance: DeployStepOne = component.instance() as DeployStepOne;

            const dropdownEvent: any = {
                selectedItem: `${packageOne.name}@${packageOne.version} (packaged)`
            };

            instance.selectPackage(dropdownEvent);

            onPackageChangeStub.should.have.been.calledOnceWithExactly(packageOne);
        });

        it('should be able to select a package without a version', () => {
            const component: ReactWrapper<DeployStepOne> = mount(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={packageOne} selectedWorkspace={undefined} workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            const instance: DeployStepOne = component.instance() as DeployStepOne;

            const dropdownEvent: any = {
                selectedItem: `${packageThree.name} (packaged)`
            };

            instance.selectPackage(dropdownEvent);

            onPackageChangeStub.should.have.been.calledOnceWithExactly(packageThree);
        });

        it('should be able to select a workspace', () => {
            const component: ReactWrapper<DeployStepOne> = mount(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={undefined} selectedWorkspace='workspaceOne' workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            const instance: DeployStepOne = component.instance() as DeployStepOne;

            const dropdownEvent: any = {
                selectedItem: `workspaceOne (open project)`
            };

            instance.selectPackage(dropdownEvent);

            onPackageChangeStub.should.have.been.calledOnceWithExactly(undefined, 'workspaceOne');
        });
    });

    describe('packageWorkspace', () => {
        it('should package workspace', () => {
            const component: ReactWrapper<DeployStepOne> = mount(<DeployStepOne packageEntries={[packageOne, packageTwo, packageThree]} selectedPackage={undefined} selectedWorkspace='workspaceOne' workspaceNames={['workspaceOne']} onPackageWorkspace={onPackageWorkspaceStub} deletedSelectedPackage={false} onPackageChange={onPackageChangeStub} />);
            const instance: DeployStepOne = component.instance() as DeployStepOne;

            instance.packageWorkspace();

            onPackageWorkspaceStub.should.have.been.calledOnceWithExactly('workspaceOne');
        });
    });

});
