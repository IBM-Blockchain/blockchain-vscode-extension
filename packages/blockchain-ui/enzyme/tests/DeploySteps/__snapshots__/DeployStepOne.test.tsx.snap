// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeployStepOne component render should render the expected snapshot 1`] = `
Array [
  <div
    className="bx--row margin-bottom-10"
  >
    <div
      className="bx--col"
    >
      <div
        style={
          Object {
            "width": 384,
          }
        }
      >
        <div
          className="bx--dropdown__wrapper bx--list-box__wrapper"
        >
          <span
            className="bx--label"
          >
            Choose a smart contract to deploy
          </span>
          <div
            aria-label="dropdown"
            className="bx--dropdown bx--list-box"
            id="package-select"
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            <div
              aria-controls={null}
              aria-disabled={false}
              aria-expanded={false}
              aria-haspopup={true}
              aria-labelledby="dropdown-label-package-select dropdown-field-label-package-select"
              aria-owns={null}
              className="bx--list-box__field"
              data-toggle={true}
              onBlur={[Function]}
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex="0"
              type="button"
            >
              <span
                className="bx--list-box__label"
                htmlFor="downshift-0-input"
                id="dropdown-field-label-package-select"
              >
                mycontract@0.0.1 (packaged)
              </span>
              <div
                className="bx--list-box__menu-icon"
              >
                <svg
                  aria-label="Open menu"
                  focusable="false"
                  height={16}
                  name="chevron--down"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  viewBox="0 0 16 16"
                  width={16}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    points="8,11 3,6 3.7,5.3 8,9.6 12.3,5.3 13,6"
                  />
                  <title>
                    Open menu
                  </title>
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="bx--row"
  >
    <div
      className="bx--col-lg-10"
    >
      <ul
        className="bx--accordion bx--accordion--end"
      >
        <li
          className="bx--accordion__item"
          onAnimationEnd={[Function]}
        >
          <button
            aria-expanded={false}
            className="bx--accordion__heading"
            onClick={[Function]}
            onKeyDown={[Function]}
            title="Expand/Collapse"
            type="button"
          >
            <svg
              aria-label="Expand/Collapse"
              className="bx--accordion__arrow"
              focusable="false"
              height={16}
              preserveAspectRatio="xMidYMid meet"
              role="img"
              viewBox="0 0 16 16"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="11,8 6,13 5.3,12.3 9.6,8 5.3,3.7 6,3"
              />
            </svg>
            <div
              className="bx--accordion__title"
            >
              How does Fabric v2.X smart contract deployment work?
            </div>
          </button>
          <div
            className="bx--accordion__content"
          >
            <p>
              TODO
            </p>
          </div>
        </li>
      </ul>
    </div>
  </div>,
]
`;
